### PR TITLE
Mustafa's Authorization and Teams API Routes

### PR DESCRIPTION
**Pull request by id-mustafa**
Hello!
Since I am using part of Nick's work in this PR. I'd like to give him credit for setting up the team service and team model. He is also attached to this PR.  @NickA02 

This PR houses the branch with all the changes regarding generating a JWT and the Team routes. Here are the following file changes with a brief description of their intended purposes:

backend/api/auth.py
- This houses the API that creates a JWT through a post request that passes in the name and password in the request body. You can also grab the currently authenticated user, and logout the authenticated user.  

backend/api/team.py
- This houses all the APIs for the teams. You can retrieve a team by their name and by their id. You can also retrieve all teams.

backend/models/auth.py
- This houses all the models pertaining to the Token, TokenData, and LoginData.

backend/services/auth.py
- Created the service methods needed to verify, authenticate, and create/decode a JWT.

backend/services/team.py
- Created 2 methods on top of Nick's previous work. The get_team() and get_all_teams() methods for the backend Teams api's to work.

fake_data/team.py
- Created the test data for the team.py. What I did here was just ensure that I had a sample of data to test with. I also tested out the functions created by Nick and they work beautifully (Thank you Nick!). 

backend/.env.development (Needs to be configured per contributor)
- This .env file houses the .env needed for the JWT to work correctly. You set an algorithm, a secret key that was generated with the following command -- `openssl rand -hex 32` and lastly the expiration time. I currently set mine to be at 30 minutes.

I hope these changes are comprehensive at what I did to build out these routes. Please let me know how you feel about them. Thank you!

### CODE REVIEW

**Comment by ItIsAndrewL** [comment]
> backend/models/auth.py
> 
> * This houses all the models pertaining to the LoginData. I had to list these as standard Pydantic Models instead of SQLModels since we didn't wanna create a whole seperate table for this.

No different than what you did here as far as I know but SQLModel will only create a table if you pass in `table=true` to the class definition if you find there to be some sort of difference.

File: README.md [hunk 1]
```diff
@@ -31,7 +33,17 @@ Run the following commands:
 
 ### Backend
 
-To help with backend implementation, you can reference the `count` demo feature. 
+To help with backend implementation, you can reference the `count` demo feature.
+
+#### Authentication - JWT
+
+Ensure that you have these packages installed: `pip3 install pyjwt` to ensure your project works well with the authorization system. Navigate to your `backend` folder and create yourself a `.env.development`. You will need to run the following command to generate yourself a secret key: `openssl rand -hex 32`. To confirm that the route works as expected, clear your database and run the test data generated for the `teams.py` test data.
```
Comment (ItIsAndrewL) [inline README.md:40]: Can you add this to the `requirements.txt` so we can just get it in the install steps?

File: README.md [hunk 2]
```diff
@@ -31,7 +33,17 @@ Run the following commands:
...
-To help with backend implementation, you can reference the `count` demo feature. 
+To help with backend implementation, you can reference the `count` demo feature.
+
+#### Authentication - JWT
+
+Ensure that you have these packages installed: `pip3 install pyjwt` to ensure your project works well with the authorization system. Navigate to your `backend` folder and create yourself a `.env.development`. You will need to run the following command to generate yourself a secret key: `openssl rand -hex 32`. To confirm that the route works as expected, clear your database and run the test data generated for the `teams.py` test data.
+
+The format for the `.env.development` file is:
+
+SECRET_KEY=1e2b4e4140036f4ed0131bd6d3a2061181735211c2364ccf47ac410a98c03682
```
Comment (ItIsAndrewL) [inline README.md:44]: Maybe take out the pre-generated key here so we're not tempted to just re-use this one.

File: backend/api/auth.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+@api.post("/login", response_model=Token, tags=["Auth"])
+async def authenticate(
+    login_data: LoginData,
+    auth_service: AuthService = Depends(),
+    team_service: TeamService = Depends(),
+):
+    """Authenticates a team member and returns a JWT Token."""
+    team = team_service.get_team(login_data.name)
+    if not team:
+        raise HTTPException(
```
Comment (ItIsAndrewL) [inline backend/api/auth.py:26]: A more idiomatic way to handle this is to create a custom exception in `services/exceptions.py` and define a custom exception handler in `main.py` (see the comments) so you can simply raise an exception in the service and have FastAPI take care of raising the HTTPError.

File: backend/api/auth.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Team not found",
+        )
+
+    token = auth_service.authenticate_and_generate_token(
+        name=login_data.name, password=login_data.password, team=team
+    )
+    if not token:
+        raise HTTPException(
```
Comment (ItIsAndrewL) [inline backend/api/auth.py:35]: See other comment.

File: backend/config.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from dotenv import load_dotenv
```
Comment (ItIsAndrewL) [inline backend/config.py:1]: Did you add `dotenv` to the `requirements.txt`?

File: backend/api/team.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@api.get("/", tags=["Teams"])
+async def get_teams(team_service: TeamService = Depends()):
+    """Gets all teams"""
+    return team_service.get_all_teams()
+
+
+@api.get("/id/{id}", tags=["Teams"])
+async def get_team(id: int, team_service: TeamService = Depends()) -> Team:
```
Comment (ItIsAndrewL) [inline backend/api/team.py:22]: Question: I see these are currently just test routes, but we should probably have a function that returns the team object of the currently logged in user. How would we go about reading the JWT that is sent and retrieving that team object?

Comment (ItIsAndrewL) [reply backend/api/team.py:22]: I can create this, but just need the info!

File: backend/models/team.py
Comment (ItIsAndrewL) [inline backend/models/team.py:1]: Think we should have Nick submit his PR first so he gets credit for these changes?

File: frontend/src/App.tsx
Comment (ItIsAndrewL) [inline frontend/src/App.tsx:1]: Looks like you'll need to fix these merge conflicts as these frontend changes shouldn't be showing up.

**Review by ItIsAndrewL — CHANGES_REQUESTED** [review]
Hi @id-mustafa this looks great so far! Looks like you are missing the middleware functionality we discussed to have any request be able to dependency inject a function which reads the JWT token and responds with that associated Team or throws an access error.

File: README.md (hunk 1 above)
Comment (id-mustafa) [reply README.md:40]: Should already be added, I froze my dependencies into requirements.txt

File: backend/config.py (hunk 5 above)
Comment (id-mustafa) [reply backend/config.py:1]: froze the dotenv to the requirements.txt file

File: README.md (hunk 2 above)
Comment (id-mustafa) [reply README.md:44]: Done

File: backend/api/auth.py (hunk 3 above)
Comment (id-mustafa) [reply backend/api/auth.py:26]: I added that in🫡

File: backend/api/auth.py (hunk 4 above)
Comment (id-mustafa) [reply backend/api/auth.py:35]: Done

File: backend/api/team.py (hunk 6 above)
Comment (id-mustafa) [reply backend/api/team.py:22]: Returning the team object might compromise the safety of the passwords. Instead, I encoded TokenData and created a route to decode that data back when the user is actively logged in🫡.

File: backend/models/team.py
Comment (id-mustafa) [reply backend/models/team.py:1]: I made a flagging mechanism here. Don't know how that might affect his changes. Can I list him as a co-author?

Comment (ItIsAndrewL) [reply backend/models/team.py:1]: Should work

**Review by ItIsAndrewL — APPROVED** [review]
Okay yall this looks good now. Lets move on to more of the meat and potatoes. I do have some concerns with the `active_JWT` field since we can have multiple people logged into one account at once, which would break this. I'll move this to a ticket and we can talk about it later.

**Pull request merged**